### PR TITLE
Add [Jenkins] Jacoco coverage badge ☘

### PR DESCRIPTION
**Pull request by stanleyhlng**
This PR added a new badge support for Jenkins Jacoco Code Coverage using `/jenkins/j` in the route.

> https://img.shields.io/jenkins/j/https/[jenkins-project-url].svg

Url Pattern for Jenkins Jacoco coverage report in json format.
> https://[jenkins-host]/job/[project]/job/[branch]/lastBuild/jacoco/api/json

Sample Jenkins Jacoco coverage report in json format.
```
{
   "_class":"hudson.plugins.jacoco.report.CoverageReport",
   "branchCoverage":{
      "covered":1796,
      "missed":819,
      "percentage":69,
      "percentageFloat":68.68069,
      "total":2615
   },
   "classCoverage":{
      "covered":235,
      "missed":48,
      "percentage":83,
      "percentageFloat":83.038864,
      "total":283
   },
   "complexityScore":{
      "covered":2398,
      "missed":1205,
      "percentage":67,
      "percentageFloat":66.55565,
      "total":3603
   },
   "instructionCoverage":{
      "covered":39498,
      "missed":9508,
      "percentage":81,
      "percentageFloat":80.5983,
      "total":49006
   },
   "lineCoverage":{
      "covered":7724,
      "missed":1723,
      "percentage":82,
      "percentageFloat":81.761406,
      "total":9447
   },
   "methodCoverage":{
      "covered":1627,
      "missed":597,
      "percentage":73,
      "percentageFloat":73.15648,
      "total":2224
   },
   "previousResult":{
      "_class":"hudson.plugins.jacoco.report.CoverageReport"
   }
}
```

### CODE REVIEW

**Comment by paulmelnikow** [comment]
Hi! Thanks for this! Could you add tests for this service? We have [a great tutorial](https://github.com/badges/shields/blob/master/doc/service-tests.md). If you have any questions along the way, feel free to post here and one of the maintainers will respond.

**Comment by paulmelnikow** [comment]
Hi @stanleyhlng, just thought I would follow up! Would you be able to add tests for the new badge?

**Comment by stanleyhlng** [comment]
@paulmelnikow yes I'm gonna add tests and revise my pr. stay tuned.

**Comment by stanleyhlng** [comment]
@paulmelnikow I have added tests for the jenkins jacoco coverage service test.  please feel free to view and give me some feedback.

console log:
```
➜  shields git:(jenkins-coverage-jacoco) npm run test:services -- --only=jenkins

> gh-badges@1.3.0 test:services /Users/sng/Projects/stanleyhlng/github/shields
> mocha --delay lib/service-test-runner/cli.js "--only=jenkins"

Running tests for 1 services: jenkins.



0402172417 Server is starting up: http://localhost:1111/
No fallback font set.
Unable to load fallback font. Using Helvetica-Bold instead.
  Jenkins
    latest version
      ✓
	[ GET http://localhost:1111/jenkins/plugin/v/blueocean.json ]
    version 0
      ✓
	[ GET http://localhost:1111/jenkins/plugin/v/blueocean.json ]
    inexistent artifact
      ✓
	[ GET http://localhost:1111/jenkins/plugin/v/inexistent-artifact-id.json ]
    connection error
      ✓
	[ GET http://localhost:1111/jenkins/plugin/v/blueocean.json ]
    jacoco: inaccessible | request error
      ✓
	[ GET http://localhost:1111/jenkins/j/https/updates.jenkins-ci.org/job/hello-project/job/master.json ]
    jacoco: inaccessible | invalid coverage object
      ✓
	[ GET http://localhost:1111/jenkins/j/https/updates.jenkins-ci.org/job/hello-project/job/master.json ]
    jacoco: unknown | invalid coverage (non-numeric)
      ✓
	[ GET http://localhost:1111/jenkins/j/https/updates.jenkins-ci.org/job/hello-project/job/master.json ]
    jacoco: 81% | valid coverage
      ✓
	[ GET http://localhost:1111/jenkins/j/https/updates.jenkins-ci.org/job/hello-project/job/master.json ]
    jacoco: unknown | exception handling
      ✓
	[ GET http://localhost:1111/jenkins/j/https/updates.jenkins-ci.org/job/hello-project/job/master.json ]


  9 passing (618ms)
```

coverage report:  
screen shot 2018-04-02 at 10 19 58 am


Thanks for the service test tutorial guideline.  It was very helpful  🎉

File: server.js [hunk 1]
```diff
@@ -4877,6 +4877,61 @@ cache(function(data, match, sendBadge, request) {
   });
 }));
 
+// Jenkins coverage integration (jacoco)
+camp.route(/^\/jenkins(?:-ci)?\/j\/(http(?:s)?)\/([^/]+)\/(.+)\.(svg|png|gif|jpg|json)$/,
+cache(function(data, match, sendBadge, request) {
+  var scheme = match[1];  // http(s)
```
Comment (PyvesB) [inline server.js]: I suggest using the ES6 `let` and `const` keywords to declare variables. We still have a lot of code exclusively making use of `var`, but we try to update things as we go along. 😉

File: server.js [hunk 2]
```diff
@@ -4877,6 +4877,61 @@ cache(function(data, match, sendBadge, request) {
...
+  var host = match[2];  // example.org:8080
+  var job = match[3];  // folder/job
+  var format = match[4];
+  var options = {
+    json: true,
+    uri: scheme + '://' + host + '/job/' + job
```
Comment (PyvesB) [inline server.js]: Would this be a good candidate for a template string?

File: server.js [hunk 3]
```diff
@@ -4877,6 +4877,61 @@ cache(function(data, match, sendBadge, request) {
...
+  if (serverSecrets && serverSecrets.jenkins_user) {
+    options.auth = {
+      user: serverSecrets.jenkins_user,
+      pass: serverSecrets.jenkins_pass
+    };
+  }
+
+  var badgeData = getBadgeData('coverage', data);
+  request(options, function(err, res, json) {
+    if (err !== null) {
```
Comment (PyvesB) [inline server.js]: We've now got a slightly more comprehensive `checkErrorResponse` helper function to cover error cases. Quite a lot of the newer badge implementations use it if you want to get an example of how it works.

File: server.js [hunk 4]
```diff
@@ -4877,6 +4877,61 @@ cache(function(data, match, sendBadge, request) {
...
+
+    try {
+      var coverageObject = json.instructionCoverage;
+      if (coverageObject === undefined) {
+        badgeData.text[1] = 'inaccessible';
+        sendBadge(format, badgeData);
+        return;
+      }
+      var coverage = coverageObject.percentage;
+      if (+coverage !== +coverage) {
```
Comment (PyvesB) [inline server.js]: I'm no Javascript expert, but wouldn't is be more readable to use `isNaN` for this check?

File: services/jenkins/jenkins.tester.js [hunk 5]
```diff
@@ -40,3 +40,40 @@ t.create('connection error')
   .get('/plugin/v/blueocean.json')
   .networkOff()
   .expectJSON({ name: 'plugin', value: 'inaccessible' });
+
```
Comment (PyvesB) [inline services/jenkins/jenkins.tester.js]: Could you please specify "cobertura: " at the beginning of the previous test names?

File: services/jenkins/jenkins.tester.js [hunk 6]
```diff
@@ -40,3 +40,40 @@ t.create('connection error')
...
+t.create('jacoco: unknown | invalid coverage (non-numeric)')
+  .get('/j/https/updates.jenkins-ci.org/job/hello-project/job/master.json')
+  .intercept(nock => nock('https://updates.jenkins-ci.org')
+    .get('/job/hello-project/job/master/lastBuild/jacoco/api/json?tree=instructionCoverage[covered,missed,percentage,total]')
+    .reply(200, { instructionCoverage: { covered: 39498, missed: 9508, percentage: 'non-numeric', percentageFloat: 80.5983, total: 49006 } })
+  )
+  .expectJSONTypes({ name: 'coverage', value: 'unknown' });
+
+t.create('jacoco: 81% | valid coverage')
```
Comment (PyvesB) [inline services/jenkins/jenkins.tester.js]: Shall we put this happy test case as the first jacoco one and leave the error cases grouped together?

**Comment by RedSparr0w** [comment]
Thanks for this PR!

I've just done a quick [diff check](https://www.diffchecker.com/6M2VtpSs) between the `cobertura` & `jacoco` implementations, and it seems there is only a couple of lines different, I think it would be best to merge these 2 into 1.

quick example with some of the possible required changes:
```javascript
// Jenkins coverage integration (cobertura & jococo)
camp.route(/^\/jenkins(?:-ci)?\/(c|j)\/(http(?:s)?)\/([^/]+)\/(.+)\.(svg|png|gif|jpg|json)$/,
cache(function(data, match, sendBadge, request) {
  var implementation_type = math[1] // cobertura/jacoco
  var scheme = match[2];  // http(s)
  var host = match[3];  // example.org:8080
  var job = match[4];  // folder/job
  var format = match[5];

  ...

  var options = {
    json: true,
    uri: scheme + '://' + host + '/job/' + job;
  };
  if (job.indexOf('/') > -1 ) {
    options.uri = scheme + '://' + host + '/' + job;
  }

  switch (implementation_type){
    case 'c':
      options.uri += '/lastBuild/cobertura/api/json?tree=results[elements[name,denominator,numerator,ratio]]';
      break;
    case 'j':
      options.uri += '/lastBuild/jacoco/api/json?tree=instructionCoverage[covered,missed,percentage,total]';
      break;
  }

  ...
```

**Comment by stanleyhlng** [comment]
@PyvesB @RedSparr0w I have revised my PR based on your feedback.  Please take your time to review.  Thanks!

**Comment by PyvesB** [comment]
@RedSparr0w have you had a chance to look at this? 😃

**Comment by stanleyhlng** [comment]
@RedSparr0w I have triggered the service test by pushing empty commit.  All tests are passed.

**Comment by RedSparr0w** [comment]
Thanks, Merged!
😄

**Pull request merged**